### PR TITLE
Implement little endian ascending transaction ordering (CTOR)

### PR DESCRIPTION
**Pull request by gandrewstone**
Uses a sequential validation algorithm.

### CODE REVIEW

File: src/validation/validation.cpp [hunk 1]
```diff
@@ -2454,6 +2744,11 @@ void UpdateTip(CBlockIndex *pindexNew)
         // alert if significant signaling changes.
         CheckAndAlertUnknownVersionbits(chainParams, chainActive.Tip());
     }
+
+    if (IsNov152018Enabled(chainParams.GetConsensus(), pindexNew))
+    {
+        enableCanonicalTxOrder = true;
+    }
```
Comment (ptschip) [inline src/validation/validation.cpp:2792]: Once it's on , it's on, but what if we have a chain reorg just after activation?  Also, does this really need to be a tweak, or was it for testing purposes only...I think it would be better to drop the tweak, but at minimum you'd need to reset to false if the fork becomes disabled.

File: src/globals.cpp [hunk 2]
```diff
@@ -305,6 +304,13 @@ CTweakRef<bool> enableDataSigVerifyTweak("consensus.enableDataSigVerify",
     "true if OP_DATASIGVERIFY is enabled.",
     &enableDataSigVerify);
 
+CTweak<bool> enableCanonicalTxOrder("consensus.enableCanonicalTxOrder",
+    "True if canonical transaction ordering is enabled.  Reflects the actual state so may be switched on or off by"
+    " fork time flags and blockchain reorganizations.",
+    false);
+
+std::atomic<bool> ltorEnabled{false}; // Is lexical transaction ordering enabled?
+
```
Comment (ptschip) [inline src/globals.cpp]: do we still need ltorEnabled if we're using the tweak?

File: src/coins.cpp [hunk 3]
```diff
@@ -577,6 +577,12 @@ void UpdateCoins(const CTransaction &tx, CValidationState &state, CCoinsViewCach
             inputs.SpendCoin(txin.prevout, &txundo.vprevout.back());
         }
     }
+}
+
+void UpdateCoins(const CTransaction &tx, CValidationState &state, CCoinsViewCache &inputs, CTxUndo &txundo, int nHeight)
+{
+    // mark inputs spent
+    SpendCoins(tx, state, inputs, txundo, nHeight);
     // add outputs
```
Comment (ptschip) [inline src/coins.cpp:586]: I don't understand why this is part of CTOR? why would we spend coins when updating coins or is this for some other bug fix?

File: qa/rpc-tests/mempool_accept.py [hunk 4]
```diff
@@ -439,7 +439,7 @@ def Test():
     t = MyTest(True)
     bitcoinConf = {
         "debug": ["blk", "mempool", "net", "req"],
-        "blockprioritysize": 2000000,  # we don't want any transactions rejected due to insufficient fees...
+        "blockminsize": 32000000,  # we don't want any transactions rejected due to insufficient fees...
```
Comment (sickpig) [inline qa/rpc-tests/mempool_accept.py]: why is this change related to ctor implementations?

File: src/coins.cpp (hunk 3 above)
Comment (sickpig) [reply src/coins.cpp:586]: seconded

File: src/validation/validation.cpp [hunk 5]
```diff
@@ -2161,33 +2140,243 @@ bool ConnectBlock(const CBlock &block,
...
+    bool fParallel,
+    bool fScriptChecks,
+    CAmount &nFees,
+    CBlockUndo &blockundo,
+    std::vector<std::pair<uint256, CDiskTxPos> > &vPos,
+    std::vector<uint256> &vHashesToDelete)
+{
+    nFees = 0;
+    int64_t nTime2 = GetTimeMicros();
+    LOGA("Canonical ordering for %s\n", block.GetHash().ToString());
```
Comment (sickpig) [inline src/validation/validation.cpp]: I'd rather use BLK category here.

File: src/globals.cpp (hunk 2 above)
Comment (sickpig) [reply src/globals.cpp]: seems to me that `ltorEnabled` is not used at all in this changeset.

File: src/miner.cpp [hunk 6]
```diff
@@ -228,6 +234,13 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript &sc
         LOGA("CreateNewBlock(): total size %llu txs: %llu fees: %lld sigops %u\n", nBlockSize, nBlockTx, nFees,
             nBlockSigOps);
 
+        // sort tx if there are any and the feature is enabled
+        if (enableCanonicalTxOrder.Value() && pblock->vtx.size() > 1)
```
Comment (sickpig) [inline src/miner.cpp]: I know this is WIP but was wondering why you don't hook up CTOR validation to nov 15 2018 activation code via helpers like `IsNov152018Enabled(Params().GetConsensus(), chainActive.Tip()))`

File: src/tweak.h [hunk 7]
```diff
@@ -176,6 +176,14 @@ class CTweakRef : public CTweakBase
...
         return *value;
     }
+
+    CTweakRef &operator=(const DataType &d)
+    {
+        std::lock_guard<std::mutex> lck(cs_tweak);
+        if (value)
+            *value = d;
+        return *this;
+    }
```
Comment (sickpig) [inline src/tweak.h:186]: what's the purpose of the above change?

File: src/tweak.h [hunk 8]
```diff
@@ -228,6 +236,13 @@ class CTweak : public CTweakBase
         std::lock_guard<std::mutex> lck(cs_tweak);
         return value;
     }
+
+    CTweak &operator=(const DataType &d)
+    {
+        std::lock_guard<std::mutex> lck(cs_tweak);
+        value = d;
+        return *this;
+    }
```
Comment (sickpig) [inline src/tweak.h:245]: ditto

File: src/txadmission.cpp [hunk 9]
```diff
@@ -344,7 +345,12 @@ void ThreadTxAdmission()
                     // and commitment will not be clean
                     CCriticalBlock lock(csTxInQ, "csTxInQ", __FILE__, __LINE__);
                     if (txInQ.empty())
+                    {
+                        // speed up tx chunk processing when there is nothing else to do
+                        if (acceptedSomething)
+                            cvCommitQ.notify_all();
```
Comment (sickpig) [inline src/txadmission.cpp:356]: Isn't this change orthogonal to CTOR?

File: src/validation/validation.cpp (hunk 1 above)
Comment (sickpig) [reply src/validation/validation.cpp:2792]: agreed

File: src/miner.cpp (hunk 6 above)
Comment (gandrewstone) [reply src/miner.cpp]: It is hooked up.  The reason for the tweak is to allow manual override as per BUIP098

File: src/tweak.h (hunk 7 above)
Comment (gandrewstone) [reply src/tweak.h:186]: allows code to set the tweak value.  Previously the member variable was public.  When it was moved to be private, a getter was implemented but not a setter.

File: src/txadmission.cpp (hunk 9 above)
Comment (gandrewstone) [reply src/txadmission.cpp:356]: it could be factored out, but is needed to speed up the ctor.py test since that test creates a lot of dependent tx and without this change the tx admission pipeline only handles 1 dependent tx every 2 seconds.

File: src/validation/validation.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/validation/validation.cpp]: yes

File: qa/rpc-tests/mempool_accept.py (hunk 4 above)
Comment (gandrewstone) [reply qa/rpc-tests/mempool_accept.py]: its not, I will revert

File: src/coins.cpp (hunk 3 above)
Comment (gandrewstone) [reply src/coins.cpp:586]: This is a refactor with no semantic changes.  UpdateCoins needs to "spend" the inputs of the tx, and add the outputs of the tx into the UTXO set.  But the OTI algorithm needs to do each step separately and in the reverse order (first it AddCoins and second it SpendCoins).  Therefore I took the original code and factored it into a new function called "SpendCoins" and call that function here.

File: src/globals.cpp (hunk 2 above)
Comment (gandrewstone) [reply src/globals.cpp]: removed.  This was what I used before the tweak.

File: src/validation/validation.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/validation/validation.cpp:2792]: It is very useful as a tweak for testing as you suggest.  However, its also a tweak to match the BUIP098's requirement to allow a manual override.

Comment (sickpig) [reply src/validation/validation.cpp:2792]: What if Nov fork get "reorged" and node operator won't switch this tweak manually?

**Comment by ptschip** [comment]
but still i think you need the following in the event of a deeper reorg
just after the fork has activated?

+else
+{
+enableCanonicalTxOrder = false;
+}

File: src/validation/validation.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/validation/validation.cpp:2792]: What I am proposing is if the Nov fork time is "on" (set to non-zero) then it sets this tweak to true or false during blockchain extensions or reorgs.  If the operator sets the nov fork time tweak to 0, then his settings for the "feature" tweaks won't be overridden.  If the operator has manually set it to "on", then in practice we will be too far beyond the fork point for a reorg to matter.

**Comment by gandrewstone** [comment]
@ptschip yes absolutely need to handle and test the reorg case.  This and a few other things is why the PR is still set as WIP.  But its nice to be able to run travis and also check the blue-sky case against the forked testnet3

**Comment by ptschip** [comment]
BTW, you need to add  ctor.py to rpc-tests.py

File: src/txadmission.cpp (hunk 9 above)
Comment (ptschip) [inline src/txadmission.cpp:356]: Even though currently there is only one thread for the commitq shouldn't this really be:

-cvCommitQ.notify_all();
+cvCommitQ.notify_one();

Comment (gandrewstone) [reply src/txadmission.cpp:356]: As you say, it doesn't really matter in this case.  But generally you want to "notify" as many times as items you have to process because you don't know how many processing threads there are, and error on the side of over-notifying because worst case in over-notifying is a thread wakes and then sleeps right away.  But worst case in the under-notifying case is a thread doesn't wake up so processing is delayed.  So in this case I have potentially accepted many tx so I notify all.

**Comment by ptschip** [comment]
needs a rebase

File: src/validation/validation.cpp [hunk 10]
```diff
@@ -2161,33 +2154,248 @@ bool ConnectBlock(const CBlock &block,
...
+
+    // Begin Section for Boost Scope Guard
+    {
+        // Scope guard to make sure cs_main is set and resources released if we encounter an exception.
+        BOOST_SCOPE_EXIT(&fParallel) { PV->SetLocks(fParallel); }
+        BOOST_SCOPE_EXIT_END
+
+        // Outputs then Inputs algorithm: add outputs to the coin cache
+        // and validate lexical ordering
+        uint256 prevTxHash;
```
Comment (ptschip) [inline src/validation/validation.cpp:2242]: Don't we just need this "add outputs first" loop in ConnectBlock() when the fork is activated and not this entire code block, seems like a lot of duplicate code. 

Other than this last comment I'm done my review and the tests seem reliable now.

**Review by ptschip — APPROVED** [review]
Looks good, except maybe we need to remove some code dup, and the tests are running reliably now.
